### PR TITLE
SMBus :expanded SoloBMS to 6s - providing support for madhacker BMSOne

### PR DESCRIPTION
**Pull request by AndKe**
Added support for extended cell count.
This change is perfectly backwards compatible with 3DR Batteries as well as BMSone.
The addition enables the BMSOne to monitor up to six cells.

### CODE REVIEW

**Comment by peterbarker** [comment]
> Added support for extended cell count. This change is perfectly backwards compatible with 3DR Batteries as well as BMSone. The addition enables the BMSOne to monitor up to six cells.

I did not see this coming :-)

Have you actually tested this patch on a Solo with its normal battery setup?

**Comment by AndKe** [comment]
Yes, it is tested and does not affect the standard battery. 
More than that: it's verified using a logic analyzer and the BMS datasheet that the Solo batteries do not care about 0x29 - it's not an undocumented command or anything like that.

**Comment by proficnc** [comment]
This looks great! Good to have some new battery options!

**Comment by AndKe** [comment]
bump ... can I have this merged please - it works for lots of people(with Solo's).

**Comment by AndKe** [comment]
@peterbarker  - please merge it.

**Comment by AndKe** [comment]
@tridge  Please merge?

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.h [hunk 1]
```diff
@@ -18,6 +18,7 @@ class AP_BattMonitor_SMBus_Solo : public AP_BattMonitor_SMBus
     void timer(void) override;
 
     uint8_t _button_press_count;
+    bool _use_extended = false;
```
Comment (IamPete1) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.h:21]:
```suggestion
    bool _use_extended;
```
Don't need to initialize false.

File: libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp [hunk 2]
```diff
@@ -39,12 +41,12 @@ AP_BattMonitor_SMBus_Solo::AP_BattMonitor_SMBus_Solo(AP_BattMonitor &mon,
...
 void AP_BattMonitor_SMBus_Solo::timer()
 {
-    uint8_t buff[8];
+    uint8_t buff[12];
     uint32_t tnow = AP_HAL::micros();
 
 
     // read cell voltages
-    if (read_block(BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE, buff, 8)) {
+    if (!_use_extended and read_block(BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE, buff, 8)) {
```
Comment (IamPete1) [inline libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Solo.cpp:49]:
```suggestion
    if (!_use_extended && read_block(BATTMONITOR_SMBUS_SOLO_CELL_VOLTAGE, buff, 8)) {
```
We don't typically use the full words.

**Comment by AndKe** [comment]
@IamPete1  Thank you very much.

**Review by tridge — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks @AndKe !

**Pull request merged**